usr/src/lib/libzfs/common/libzfs_dataset.c

```diff
@@ -854,16 +858,18 @@ libzfs_mnttab_find(libzfs_handle_t *hdl, const char *fsname,
 			return (ENOENT);
 	}
 
+	(void) mutex_lock(&hdl->libzfs_mnttab_cache_lock);
```

Comment:
Use mutex_enter(), the version in user land. It will do all the error checking and abort appropriately. Otherwise, you need to check the return value and blow up if it's non-zero so we abort on deadlock.